core/node/node_framework/src/service/mod.rs

```diff
@@ -145,30 +148,43 @@ impl ZkStackService {
             .collect();
 
         // Run the tasks until one of them exits.
-        // TODO (QIT-24): wrap every task into a timeout to prevent hanging.
-        let (resolved, idx, remaining) = self
+        let (resolved, resolved_idx, remaining) = self
             .runtime
             .block_on(futures::future::select_all(join_handles));
-        let task_name = tasks[idx].name.clone();
+        let resolved_task_name = tasks[resolved_idx].name.clone();
         let failure = match resolved {
             Ok(Ok(())) => {
-                tracing::info!("Task {task_name} completed");
+                tracing::info!("Task {resolved_task_name} completed");
                 false
             }
             Ok(Err(err)) => {
-                tracing::error!("Task {task_name} exited with an error: {err}");
+                tracing::error!("Task {resolved_task_name} exited with an error: {err}");
                 true
             }
             Err(_) => {
-                tracing::error!("Task {task_name} panicked");
+                tracing::error!("Task {resolved_task_name} panicked");
                 true
             }
         };
 
+        let remaining_tasks_with_timeout: Vec<_> = remaining
+            .into_iter()
+            .map(|task|
+                rt_handle.spawn(async move {
+                    if tokio::time::timeout(TASK_SHUTDOWN_TIMEOUT, task).await.is_err()
+                    {
+                        tracing::warn!(
+                            "One of the tasks didn't finish in {TASK_SHUTDOWN_TIMEOUT:?} and was dropped"
+                        );
+                    }
+                })
+            ).collect();
+
         // Send stop signal to remaining tasks and wait for them to finish.
         // Given that we are shutting down, we do not really care about returned values.
         self.stop_sender.send(true).ok();
-        self.runtime.block_on(futures::future::join_all(remaining));
+        self.runtime
+            .block_on(futures::future::join_all(remaining_tasks_with_timeout));
```

Comment:
I think you can just wrap the tasks into `timeout` and join the wrapped futures here. Then you can iterate through the results to see if any of them is a timeout error. Then you can emit a log on timeouts.

Reply:
not sure I understand what you mean...

Reply:
I mean that you can
1. Wrap every remaining task into `tokio::timeout(..)`.
2. Run `join_all` here as before, but store the results, e.g. `let results = self.runtime.block_on(..)`.
3. Then `results` will have an array of `Result`s, and you can check if any of them is an `Err`, then at least one task timed out.
4. After that, you can emit a single log line stating how many tasks timed out, without mentioning their name, but also withour repeating logs.